Adversarial and Devious PageCacheRule

chrisvest opened the pull request:
Note that this will _definitely_ produce complicated merge conflicts and build failures when merged into 3.0 and beyond, so please contact me before starting that work. We may want to do a null-merge into 3.0, and re-do the changes.

lutovich committed:
Introduced adversarial page cache

Such page cache implementation is able to wrap the given page cache and
introduce some misbehaviour to exercise the user code. It can either throw
RuntimeExceptions/IOExceptions or return a page cursor with inconsistent
reads. Misbehaviour is controlled by the given Adversary.

Changed PageCacheRule to use the AdversarialPageCache.

chrisvest committed:
Make the inconsistency injection of the `PageCacheRule` more devious

Previously, the `PageCacheRule` would by default randomly decide that some reads would be inconsistent.
Then, for those inconsistent reads, the page cursor would return zeros for all read operations.

Now, we still randomly choose in the `PageCursor.next()` call if a read will be inconsistent, but instead of just returning zeros, we are now more devious:
To start with, we just delegate all read operations and let them succeed, counting them in the process.
Then, in the `shouldRetry` method, we choose an operation at random to be the start of the inconsistent read.
We then let the reader retry, and when they get to the victim operation, we start returning random data from that point on.

This surfaces a number of bugs, which have also been fixed in this commit.

chrisvest commented:
First PR build was green. Rerunning the PR build to shake out flakiness.

chrisvest commented:
Second PR build also green, Windows build started.

MishaDemianenko commented on community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialReadPageCursor.java:
```diff
@@ -22,7 +22,11 @@
+import java.util.Arrays;
+import java.util.List;
 import java.util.Objects;
+import java.util.concurrent.ThreadLocalRandom;
```
Since now inconsistency generation logic is not that plain, It would be great to have updated java doc for this class that will briefly describe inconsistency preparation/read cycle.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/LongerShortString.java:
```diff
@@ -801,9 +801,11 @@ public static String decode( PropertyBlock block )
-        return encoding;
+        if ( encodingHeader < 0 | ENCODINGS_BY_ENCODING.length <= encodingHeader )
+        {
+            return null;
```
since now we can return null here there is one more potential NPE place at line 785

chrisvest replied:
That will never happen because the decoding happens after the `PropertyBlock` data has been consistently read from the page cache. I'll add a descriptive `assert` about this, though.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -649,11 +649,14 @@ public long scanForHighId() throws IOException
             long nextPageId = storeFile.getLastPageId();
```
will just comment here, since it's in the same class.
In `rebuildIdGeneratorSlow` long[] freedBatch can by side effect of incorrect read from cursor

chrisvest replied:
No, `rebuildIdGeneratorSlow` does not suffer from this because 1) it's using a shared write lock, and 2) it resets its state and rescans the whole page within its `shouldRetry` loop.

MishaDemianenko replied:
yes agree, right

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/PropertyType.java:
```diff
@@ -415,4 +415,5 @@ public int calculateNumberOfBlocksUsed( long firstBlock )
     }
 
     public static final byte[] EMPTY_BYTE_ARRAY = new byte[0];
```
please move those 2 constants to the top

chrisvest replied:
`PropertyType` is an enum, so it's not allowed syntax.

MishaDemianenko replied:
ouch then at least closed to field definition? @ line 296

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/ShortArray.java:
```diff
@@ -767,7 +767,12 @@ public int getRequiredBits( long value )
+        ShortArray[] values = ShortArray.values();
+        if ( typeId <= 0 | typeId > values.length )
+        {
+            return null;
```
This introduce potential NPE in DynamicArrayStore;

IllegalArgumentException during PropertyConversion in PropertyConversion:103

chrisvest replied:
Looking at it again, I'm not sure this change is needed at all. All paths leading here have already successfully copied the relevant data into PropertyBlocks and byte arrays. I'll revert this change.

MishaDemianenko commented:
would also propose to update page cursor java doc and put in <b>BOLD</b> that code inside  cursor section should not throw exceptions, but otherwise follow proposed pattern.

MishaDemianenko commented:
@chrisvest i finished first round of review

chrisvest committed:
Address review issues of adversarial and devious pages cursors

This also fixes a bug in `NodeStore.loadRecord` that was affecting the PBI, where an inconsistent read would make a node record look like it was in use, and the loadRecord would fill it with garbage.
However, when the record decoding retried after the inconsistent read, and the record was discovered to NOT be in use, the garbage data and raised inUse flag would be left in the NodeRecord object.
The `NodeCountsProcessor` in PBI would then blow up on the left-over garbage data.

chrisvest commented:
@MishaDemianenko I believe I've addressed all of your comments.

MishaDemianenko commented:
@chrisvest looks fine

MishaDemianenko commented:
@chrisvest test failures related to `ShortArray.values()`
